Fe updates from UCI

mnlevy1981 opened the pull request:
This branch contains the code changes used by the UCI group; it includes variable C:N as well as new iron forcing

nicola-wiseman committed:
add variable C to N

mnlevy1981 committed:
Updates from J. K. Moore and J. Yu at UCI

mnlevy1981 committed:
Fix code consistency issues in Fortran

Cleaned up some trailing white space, converted hard tabs to soft, and added a
few missing _r8 modifiers

mnlevy1981 committed:
Clean up trailing white space in YAML

mnlevy1981 committed:
Re-enable Fe balance check, clean up code

Stand-alone build found some unused variables and imports that I removed

mnlevy1981 committed:
Merge branch 'development' into fe_updates_from_uci

mnlevy1981 committed:
Update the 4p2z settings file and input files

Also removed trailing whitespace I inadvertantly added to marbl_diagnostics_mod
in the last commit

mnlevy1981 committed:
Missing 'cocco' value in a couple of 4p2z defaults

mnlevy1981 committed:
Clean up variable P:C and N:C logic

Some settings in marbl_settings_mod.F90 only need to be defined if
lvariable_PtoC (or NtoC) is true. Also, found a couple block in
marbl_init_tracer_metadata_mod.F90 that wasn't verifying N_ind > 0 before
modifying N tracers (N_ind = 0 if lvariable_NtoC = .false.)

mnlevy1981 committed:
Update older settings files

For cesm2.0 and cesm2.1 (3p1z or 4p1z), add lvariable_NtoC but set the default
to .false.; also renamed gQfe_0 -> gQfe_max and added the non-N settings from
the settings_latest files.

mnlevy1981 committed:
Add most of the new settings to SPECRTA file

Don't know what to use for Qn_fixed, waiting to hear back from UCI and NOAA
folks... so this will still fail the init.py test (but with fewer MARBL ERROR
lines)

mnlevy1981 committed:
Add fixed_NtoC settings file for testing

Also generate the 4p2z settings file from JSON as part of testing

mnlevy1981 committed:
Remove Qn_fixed from autotroph_settings_type

After discussion with Keith L, it looks like Qn_fixed and Qn_zoo were added to
the code base so the fixed N:C configuration looked more like the fixed P:C
configuration... but it isn't necessary since we want to use the same fixed N:C
throughout the code (as opposed to fixed P:C, which really does vary from among
the different functional types even when fixed spatially / temporally)

mnlevy1981 committed:
Changes to pass call_compute_subroutines tests

Modify marbl_io_mod to initialize N tracers to C tracers * 16/112 and set
feredsedflux and feventflux = 0.
Modify marbl_diagnostics_mod.F90 to include dust%remin*dust_to_Fe in integrand,
not righthand-side of iron conservation check
Modify marbl_interior_tendency_mod.F90 to sflux_in * desportion in P_iron%remin
before computing sflux_out.

Also cleaned up a few small things in interior_tendency_mod (use unit_system
instead of assuming cgs; initialize Lig_prod and declare it as inout in some
routines)

mnlevy1981 committed:
initialize coccoN by scaling spC

no coccoC in this initial condition file, but coccoC = 0.07*spC

mnlevy1981 committed:
Merge tag 'marbl0.48.2' into fe_updates_from_uci

mnlevy1981 committed:
Fix units and argument order

1. Had some hard-coded units instead of using unit_system for new variables
2. Want marbl_status_log to be last argument to subroutines whenever possible

mnlevy1981 committed:
Bugfix: update operator

PON%remin is PON%sflux_in * dzr_loc + ..., not PON%sflux_in + dzr_loc + ...

mnlevy1981 committed:
Update diagnostics YAML file

CI was failing the metadata check tests; now those tests should pass

mnlevy1981 committed:
Fix diagnostic for fixed N:C

If running fixed N:C, autotrophs don't have N tracer

mnlevy1981 committed:
Add dependency on base_bio_on to diagnostics

Some new variables added to diagnostics_latest.yaml are only available in runs
where the base bio tracers are active.

mnlevy1981 committed:
Fix units in some settings

There was a discrepancy in the units listed in the YAML file and the Fortran
for SiOpt, FeOpt, POPt, and NOpt, and it turned out the units were incorrect in
both places.

mnlevy1981 committed:
Move Redfield Ratios to settings file

Four of the Redfield ratios are now user-configurable, and two of those four
will have different default values when running with variable N:C instead of
fixed N:C. The other 8 ratios depend of the values of the user-configurable
ratios, and 1 of these 8 has a different formula when running with variable N:C
instead of fixed.

mnlevy1981 commented:
f9bdcd7 removes the `parameter` attribute from 12 fortran variables. Four of them are now user-configurable in the MARBL settings file, and the other eight depend on those user settings. I verified that it is bit-for-bit identical to the previous commit (faa6ce2) when running with variable N:C.

mnlevy1981 committed:
Add Redfield ratios to other settings files

Previous commit added new variables to settings_latest, this updates the rest
of the files.

mnlevy1981 committed:
VSCode pointed out duplicate code in YAML file

Block defining ABIO tracers appeared twice somehow...

mnlevy1981 committed:
More code clean-up following brief review

1. Make sure intent(in) arguments precede intent(inout) or intent(out) in all
   subroutines
2. Remove NEW CODE blocks from diagnostics_mod (where failing conservation
   tests print out each individual term in the sum); while additional output
   would be useful, we want process-based terms, not individual terms
3. Remove new lines that contain commented out variable definitions and usage
4. Revert some white-space changes surrounding do-loops, if-statements, and
   comments

mnlevy1981 committed:
One more white-space update

Trying to keep blocks within an associate statement declaration aligned

mnlevy1981 committed:
Clean up comment that refers to user_nl_pop

Comment was actually incorrect, regardless of CESM ocean component the right
file to change is user_nl_marbl

mnlevy1981 committed:
Clean up whitespace around new diagnostics

mnlevy1981 committed:
One more comment clean-up

Just a small typo in variable description

mnlevy1981 committed:
More whitespace cleanup

Aligning # in multi-line comments

mnlevy1981 committed:
Add Redfield ratio to marbl_with_4p2z settings

These two files are created with

./MARBL_generate_settings_file.py -f ../defaults/json/settings_latest+4p2z.json
     -i ../tests/input_files/settings/marbl_with_o2_consumption_scalef.settings

and

./MARBL_generate_settings_file.py -f ../defaults/json/settings_latest+4p2z.json
     -i ../tests/input_files/settings/marbl_with_o2_consumption_scalef.settings
     -u mks

mnlevy1981 committed:
Lots of src cleanup from code review

1. Previous commit swapped order of store_diagnostics_phosphorus_fluxes() and
   store_diagnostics_phosphorus_fluxes(); this commit restores order used on
   development branch
2. Better comment describing rmean_CaCO3_bury_integral
3. Consistent use of > instead of .gt. in marbl_init_tracer_metadata_mod.F90
4. Remove PON_bury_coeff from particulate_share_type
5. Only pass k level of Fefree, Fe_scavenge_rate, Fe_scavenge, and Lig_scavenge
   to compute_scavenging(); only pass k level of Fe_scavenge to
   compute_large_detritus_prod()
6. Remove comment about gQn being an output in some associate statements when
   it is treated as an input (and one of those organizes by inputs and outputs
   so I moved the "gQn =>" line up to the input section)
7. Indented an if-else block in compute_grazing()
8. Move definition of dzr_mod to right before it is first used, since that is
   where the comment describing the variable is
9. Removed two newlines that were added inadvertently
10. Removed commented block of code from compute_denitrif()
11. gQfe_max, gQp_max, and gQn_max are maximums, not initial values. Updated
    comments accordingly (only in src/, still incorrect in the settings YAML /
    JSON files -- those will get updated once we have a test to ensure
    consistency between YAML and Fortran)

mnlevy1981 commented:
ff9a042 and d335418 addressed the majority of comments from @klindsay28 during our most recent code review. Still outstanding:

1. There are inconsistencies between the units / variable long names defined in `marbl_settings_mod.F90` and those in the settings YAML files; I want to introduce a new test that flags these automatically, and will fix them once the test exists and has caught all of them
2. I want to update the forcing files to add the new reduced sediment flux and vent flux forcings, and also update the oxidized flux
3. There are scientific questions raised in the review, and I am waiting for a response from the UCI developers to an email asking how to best proceed

mnlevy1981 committed:
Update iron sediment forcing in input file

Grabbed updated iron sediment flux and new red sed / vent fluxes from POP
forcing file; also updated test suite to read the new fluxes from file

mnlevy1981 committed:
Compare settings metadata in fortran and json

Updated marbl_init_drv.F90 to add ability to write a basic YAML file with the
schema

(shortname):
  longname: (longname)
  units: (unit)

For every setting registered in marbl_settings_type. The new
settings_metadata_check.py script reads in a JSON settings file and also the
YAML file written by init.py and flags notes any of the following:

1. Variables where the longname does not match
2. Variables where the units do not match
3. Variables that are in the fortran output but not the JSON file
4. Variables that are in the JSON but not in the fortran output

run_test_suite.sh already generates settings files from every JSON file and
runs init.py; now it runs settings_metadata_check.py afterwards. It also
includes that python script when running pylint. (pylintrc was updated so it
does not report duplicate code between the two metadata_check.py scripts).

netcdf_metadata_check.py was updated because there was an indentation error at
the very bottom of the file.

mnlevy1981 committed:
Clean up units and long names for latest+4p2z

Some of the metadata in marbl_settings_mod.F90 was wrong, and some of the
changes were in the YAML / json file. I also updated the testing script itself
to account for the fact that particulate_flux_ref_depth is always provided in m
(even when running in cgs, we use it to create diags like VAR_100m and don't
want VAR_10000cm)

Also want a .gitignore file in the init test dir to ignore
settings_metadata.yaml

mnlevy1981 committed:
Clean up metadata in two settings files

Both settings_latest and settings_latest+cocco now pass the metadata
consistency check (match what is in the Fortran)

mnlevy1981 committed:
Clean up metadata for cesm2.0 and 2.1 settings

Settings for cesm2.0, cesm2.1, and cesm2.1+cocco all pass metadata consistency
check (match what is in the Fortran)

mnlevy1981 commented:
> [ff9a042](https://github.com/marbl-ecosys/MARBL/commit/ff9a042563e29121325978a61349c1d925ad28f1) and [d335418](https://github.com/marbl-ecosys/MARBL/commit/d335418461ebaa369d388ec64b2b913bcbd1a30a) addressed the majority of comments from @klindsay28 during our most recent code review. Still outstanding:
> 
> 1. There are inconsistencies between the units / variable long names defined in `marbl_settings_mod.F90` and those in the settings YAML files; I want to introduce a new test that flags these automatically, and will fix them once the test exists and has caught all of them

This has been addressed:

* cec373c added a new tool to compare Fortran metadata to YAML metadata (the `init.py` test now creates `settings_metadata.yaml`, which contains `units` and `metadata` from every available setting; `MARBL_tools/settings_metadata_check.py` compares that to the metadata that `MARBL_settings_class.py` picks up from `defaults/`) and runs that tool as part of the CI
* 37e147d, 75956a4, and ec8b709 addressed the failing CI tests
> 
> 2. I want to update the forcing files to add the new reduced sediment flux and vent flux forcings, and also update the oxidized flux

This was done in 810e465
> 
> 3. There are scientific questions raised in the review, and I am waiting for a response from the UCI developers to an email asking how to best proceed

I have a response to my email, but still need to apply the recommended fixes. Once those code mods are in and reviewed, I'll update the baselines for our test suite so we pass all the CI tests

mnlevy1981 committed:
Correct PON%sed_loss formulation

PON_bury_coeff is a scale-factor that should be applied to POC_bury_coeff
rather than treated as the actual PON burial coefficient. It was being used as
the latter in the PON%sed_loss computation, and now is properly being used as
the former

mnlevy1981 committed:
Update baselines

The variable N:C and new iron forcings both changed answers. Baselines now
include those modifications

mnlevy1981 committed:
Remove "unitless" units

Real variables have units of '1'; logicals, strings, and integers that are
tracking non-physical quantities (like index of autotroph in grazing type) have
units of ''.

The init test now writes out units : '{units}', so empty strings show up as ''
and units of 1 are treated as a string instead of integer in
settings_metadata_check.py. (Also that python file did not set error_found if
the only differences between the JSON and Fortran metadata was in the units;
that bug has been fixed).

mnlevy1981 committed:
Update test suite to use absolute paths

mnlevy1981 commented:
As of the [three most recent commits](https://github.com/marbl-ecosys/MARBL/pull/464/files/dd4600ab169be982c677f58ebdf67f1aa0c62836..a1584c029a260cd6f95e01d5827f25fc90c0e82e), this code appears to be ready to merge. I do want to run a few CESM / stand-alone MOM6 tests, but 🤞 there are no more changes to this PR before it gets merged.

mnlevy1981 committed:
Only compare settings if init PASSed

If init.py fails with a specific settings file, then we shouldn't compare the
settings metadata afterwards

mnlevy1981 committed:
Create autotroph_tracer_consistency_enforce()

This subroutine calls setup_local_tracers() [which is now the public routine
marbl_interior_tendency_setup_local_tracers()], which in turn calls
autotroph_zero_consistency_enforce() to ensure tracer values for each autotroph
are non-negative and, if any autotroph's tracer value is zero at a location
then they all are. The new subroutine then updates this%tracer with the
consistent autotroph values, leaving non-autotroph tracers unchanged.

mnlevy1981 committed:
Only set Lig_prod(k)=0 when k>kmt

When I first created a branch with the iron forcing updates, the
call_compute_subroutines stand-alone test did not run; one of the issues was
that Lig_prod(k) was not defined for k > kmt. I thought I added a line of code
to set Lig_prod(k)=0 in that case, but inadvertantly was setting Lig_prod(k)=0
everywhere (overwriting the value computed earlier in the routine when k<=kmt).
The fix is to move the Lig_prod(k)=0 line into the else block of the "if (k <=
column_kmt)" statement

mnlevy1981 committed:
Force older docutils in environment

CI tests were failing with

Could not import extension sphinx.builders.latex (exception: No module named 'docutils.utils.roman')

According to https://docutils.sourceforge.io/RELEASE-NOTES.html the
docutils/utils/roman.py file was removed in 0.22, so we require something older

mnlevy1981 committed:
Merge branch 'development' into fe_updates_from_uci

mnlevy1981 committed:
introduce dust_per_unit_fe settings variables

Rather than hard-coding values for scaling fesedflux and feventflux when
computing dust production, users can now control these via marbl_in

mnlevy1981 commented:
5745022 failed baseline comparison tests because the new `dust_per_unit_fe` parameters have different defaults than the previously-hardcoded values and also because we fixed how `dust%prod` was used in the dust flux and remin terms (`dust%hflux_in`, `dust%sflux_out`, `dust%hflux_out`, and `dust%remin` are all different). I probably should have expected to fail the `Compare settings metadata` tests as well, that will be easy to clean up but I'll wait until we have settled on final parameters before re-generating baselines.

mnlevy1981 committed:
Update settings files to include new vars

dust_per_unit_fesedflux and dust_per_unit_feventflux are listed in all settings
files; for settings_latest* the default values will match what is in the
Fortran, while they are both set to 0 for cesm2.0 and cesm2.1 settings.

mnlevy1981 committed:
Merge branch 'development' into fe_updates_from_uci

klindsay28 commented on src/marbl_interior_tendency_mod.F90:
Can we confirm that this block is consistent with mods provided by Keith M?

mnlevy1981 replied:
This line is present in the latest version of the code from UCI, although the constant in the DOC term was decreased from 2 -> 1.6 in their latest runs:

```fortran
       if ((DOC_loc(k) .gt. c0) .and. (DOCr_loc(k) .gt. c0)) then
            sinking_mass = sinking_mass * (((DOC_loc(k) * 1.6_r8) + DOCr_loc(k)) / DOCr_loc(k))
       endif
```

(cc @kristenkrumhardt)

mnlevy1981 committed:
Updated settings from UCI

mnlevy1981 committed:
Set dust%prod = 0 to prevent negative Fe

This is a temporary fix while we figure out what's happening to cause the
negative iron in the first place (it seems related to the vertical grid,
because iron typically goes negative in vanishing layers at the bottom of the
column as well as the bottommost non-negligible level; this seems to happen
when the non-negligible thickness is thinner than the thickness of the level
above it)

mnlevy1981 committed:
Zero out dust%prod differently

We have gfortran set to elevate warnings to errors, and just setting

dust%prod = c0

Leads to warnings [errors] due to unused variables being passed to
compute_scavenging(). I don't want to back out the API changes, so instead I
use

dust%prod = c0 * ([formula we want to use eventually])

Maybe compilers will be smart enough to realize it doesn't matter what they are
multiplying by 0 so they won't bother to compute it? That's doubtful, I expect
all the compilers will want to at least check to see the term on the right is a
real number instead of Inf or NaN.

mnlevy1981 committed:
update baselines for regression tests

The changes on this branch will change answers for most of the
call_compute_subroutines tests (ABIO-only does not change). We need to update
the baselines to all CI to pass.

mnlevy1981 committed:
Increase tolerance in netcdf comparison

Two tests were failing on github runners with

Variable: J_Fe
... Max relative error (1.1581517671597563e-09) exceeds 1e-09

So I set the threshold to 1e-8 instead of 1e-9

mnlevy1981 commented on MARBL_tools/netcdf_comparison.py:
```diff
@@ -30,7 +30,7 @@
 # rtol = 1e-11 fails the cgs vs mks comparison
 #              (POC_REMIN_DIC and PON_REMIN_NH4 have rel errors of ~1.4e-11)
-DEFAULT_TOLS = {'rtol' : 1e-9, 'atol' : 1e-16, 'thres' : 1e-16}
+DEFAULT_TOLS = {'rtol' : 1e-8, 'atol' : 1e-16, 'thres' : 1e-16}
```
I need to update the comment in the two lines above this (1e-9 fails the 4p2z; J_Fe has rel errors of 1.Xe-9)

mnlevy1981 commented on src/marbl_interior_tendency_mod.F90:
```diff
@@ -2675,8 +2675,12 @@ subroutine compute_scavenging(domain, k, km, marbl_tracer_indices, tracer_local,
+      ! TODO: we do not want this to be zero permanently, but it is causing issues
+      !       in MOM6 due to vanishing layers at the sea floor the non-vanishing
+      !       layer above the vanishing layers potentially being thinner than the
+      !       layer above it.
```
Update comment to mention kludgy way of setting this prevents gfortran from complaining about `fesedflux` and `feventflux` being unused

mnlevy1981 committed:
Clean-up from code review

Added some context to comments in fortran, removed misleading comment from
python
